File: Assignment1/Prob5/Prob5.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int binarysearch(int *arr, int lo, int hi, int val);
void testarr(int size, int val);
void manysearch(int *arr, int size, int val);

int main() {

	testarr(128, 5);
	testarr(512, 5);
	testarr(2048, 5);
	testarr(8192, 5);
	testarr(32768, 5);
	testarr(131072, 5);
	testarr(524288, 5);
	testarr(2097152, 5);

}

int binarysearch(int *arr, int lo, int hi, int val) {

	while (lo <= hi) {
		int mid = lo + (hi - lo)/2;

		if (arr[mid] == val) return mid;
		if (arr[mid] < val) lo = mid + 1;
		else hi = mid - 1;
	}
	return -1;
}

void testarr(int size, int val) {
	int *arr = (int*)malloc(size*sizeof(int));

	for (int i = 0; i < size; i++) {
		arr[i] = val - 1;
	}

	clock_t begin = clock();	
	manysearch(arr,size,val);
	clock_t end = clock();
	int time_elapsed = (double) (end - begin)/(CLOCKS_PER_SEC/1000);

	printf("The time it took for the array of size %d is %d milliseconds\n", size, time_elapsed);

	free(arr);
}

void manysearch(int *arr, int size, int val) {
	int bound = 10000000;

	for (int i = 0; i < bound; i++) 
		binarysearch(arr, 0, size - 1, val);
	
}

File: Assignment1/Prob5/Prob5.py
import time

def binarysearch(arr, lo, hi, val):
    while lo <= hi:

        mid = lo + (hi - lo)//2
        if arr[mid] == val:
            return mid
        elif arr[mid] < val:
            lo = mid + 1
        else:
            hi = mid - 1
    return -1

def testarr(size, val):
    arr = []

    for i in range(0,size):
        arr.append(val - 1)

    start = time.time()
    
    for i in range(0, 10000000):
        res = binarysearch(arr, 0, size-1, val)
    
    end = time.time() - start
    fin_time = end * 1000

    print("The time it took for array of size " + str(size) + " is " + str(fin_time) + " milliseconds")

def main():
    testarr(128, 5)
    testarr(512, 5)
    testarr(2048, 5)
    testarr(8192, 5)
    testarr(32768, 5)
    testarr(131072, 5)
    testarr(524288, 5)
    testarr(2097152, 5)

if __name__ == "__main__":
    main()
